### ISSUE TITLE
Add local builds

### ISSUE THREAD

**Issue by lrettig** [spacemeshos/api#55]
Rather than requiring that other projects, including go-spacemesh and the explorer backend, include this GitHub repository as a submodule and build the protobuf files as part of their build process, it seems nicer to me if we can add builds to this repository so that they can instead rely upon either 1. those builds, or better yet 2. language-specific modules that are built here. That seems better from the perspective of separation of concerns: those projects should not need to know anything about protobuf, if they're not modifying this API.

I'm investigating how to do this. Here are some options:

- just add a Makefile rule for builds
- trigger a build automatically using a git hook
- require that all PRs, or changes to hit master, include an up-to-date build
- make builds happen automatically as part of CI with every PR/commit
- automatic pre-release when changes land on master (https://github.com/marketplace/actions/automatic-releases)

CC @avive 

Task list:

- [x] golang build command (#57)
- [x] grpc-gateway build command (#57)
- [x] test if this works as a go module (https://github.com/spacemeshos/go-spacemesh/pull/2036)
- [x] make a decision about package naming
- [x] automate the build/prerelease (#62)

**Comment by lrettig** [comment]
Some options to finagle go package paths:

- move the source files to the root of this repo, and compile using `--go_opt=paths=source_relative`
- create a separate git submodule and compile into that directory (makes the build process harder since it somehow would have to push/release those changes)
- set up a go module proxy, or use something like gopkg.in
- use some fancy replace rule in `mod.go`, but I think that has to be done on side of the client that does the import
- just live with a really awkward package path, e.g., github.com/spacemeshos/api/release/go/spacemesh

**Comment by lrettig** [comment]
I added a local golang build in #57. I'm having no trouble importing it in go-spacemesh (see https://github.com/spacemeshos/go-spacemesh/commit/e8505aee7cea24406e0b3dced54441cd6532e13f). I'm going to hold off on adding automated builds until we've made a decision on package name/location since it may partially depend on that.

**Comment by lrettig** [comment]
Whoops, still have a couple of tasks left here:

- make a decision about package naming: Unless @noamnelke feels otherwise, let's just go with the "fully qualified" name for now (`github.com/spacemeshos/api/release/go/spacemesh/v1`), and leave the builds in this repository - golang, at least, makes it easy to refer to an import using a more intuitive alias
- automate the build/prerelease: will address in a separate PR

**Comment by noamnelke** [comment]
As I wrote here https://github.com/spacemeshos/api/pull/57#issuecomment-644824473 I don't think that we should include the version in the path. One downside (at least in go) of including it is that it makes upgrades more complicated (you need to change all imports everywhere instead of only `go.mod`).

**Comment by lrettig** [comment]
@noamnelke:

> As I wrote here [#57 (comment)](https://github.com/spacemeshos/api/pull/57#issuecomment-644824473) I don't think that we should include the version in the path. One downside (at least in go) of including it is that it makes upgrades more complicated (you need to change all imports everywhere instead of only `go.mod`).

Interesting. I'm following the conventions laid out in the buf lint checkers. In particular, `PACKAGE_DIRECTORY_MATCH` requires that the package live in a directory that matches it, and `PACKAGE_VERSION_SUFFIX` requires that our package name end in a version number. The logic is explained [here](https://buf.build/docs/lint-checkers#package_version_suffix): the idea is that, instead of making breaking changes to your API, you maintain the old API and make a completely new version. So I think the issue you describe, of upgrades being more complicated, is by design.

Let me know if you still disagree.

**Comment by noamnelke** [comment]
Hmm... I see where they're coming from. Interestingly they also say:

>**This is probably the most controversial checker in the `DEFAULT` category, and many users may choose to turn this off.** However, we think this is important for modern Protobuf development and highly recommend your future development abides by this checker.

I tried to think where exposing all versions concurrently would come in handy. I could come up with this - we may want to have the node expose two versions of the api at the same time (e.g. when transitioning to a new version). This would be much simplified by exposing all versions concurrently.

If this is the case, however, we'd need to think how this would be supported on the node. One solution I can think of is creating two separate grpc/json endpoints (on different ports) that each serve a different version of the api, or even simpler is to only expose one version at a time, but include both implementations and control it via a command line flag.

Bottom line - let's keep the version number in the path. I assume we'll only import it in `grpc_server.go` anyway.

**Comment by lrettig** [comment]
Thanks for looking into this. I'm glad we've thought it through.

> we'd need to think how this would be supported on the node

I agree with your second thought here:

> only expose one version at a time, but include both implementations and control it via a command line flag

I can't realistically think of a scenario where we'd want one node instance to expose multiple versions of the API.

I will keep this in mind as I add support for the new API to go-spacemesh.

**Comment by lrettig** [comment]
This was closed by #62

**Comment by noamnelke** [comment]
> I can't realistically think of a scenario where we'd want one node instance to expose multiple versions of the API.

Not saying I'm for exposing multiple versions, but still important to understand the potential benefit: if you have more than one app that consumes the api (e.g. a wallet used for managing funds, perhaps connected to your HW wallet and another for managing your smart contracts, which optimizes a different set of properties) - one app may be updated and require the newer api while the other still relies on the previous version. We don't want to force users into running two nodes in such a scenario.

While this is an edge case, if we keep both implementations in the node anyway, exposing two api servers is technically easy.

We don't need to decide in advance anyway and we should assess the situation when we get to that point.

**Comment by lrettig** [comment]
Fair point. Will keep that in mind. I agree that if it's not a huge
challenge technically or architecturally, other things being equal we may
want to allow this use case.

>
